src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -175,48 +178,96 @@ private void EnsureStartup()
+                var builder = builderFactory.CreateBuilder(_serverInstance);
+                builder.ApplicationServices = _applicationServices;
+
+                var startupFilters = _applicationServices.GetService<IEnumerable<IStartupFilter>>();
+                var configure = Startup.ConfigureDelegate;
+                foreach (var filter in startupFilters)
                 {
-                    throw new InvalidOperationException("IHostingBuilder.UseServer() is required for " + nameof(Start) + "()");
+                    configure = filter.Configure(configure);
                 }
 
-                ServerFactory = _applicationServices.GetRequiredService<IServerLoader>().LoadServerFactory(ServerFactoryLocation);
-            }
-
-            _serverInstance = ServerFactory.Initialize(_config);
-            var builderFactory = _applicationServices.GetRequiredService<IApplicationBuilderFactory>();
-            var builder = builderFactory.CreateBuilder(_serverInstance);
-            builder.ApplicationServices = _applicationServices;
+                configure(builder);
 
-            var addresses = builder.ServerFeatures?.Get<IServerAddressesFeature>()?.Addresses;
-            if (addresses != null && !addresses.IsReadOnly)
+                return builder.Build();
+            }
+            catch (Exception ex)
             {
-                var port = _config[ServerPort];
-                if (!string.IsNullOrEmpty(port))
+                if (!_captureStartupErrors)
                 {
-                    addresses.Add("http://localhost:" + port);
+                    throw;
                 }
 
-                // Provide a default address if there aren't any configured.
-                if (addresses.Count == 0)
+                // EnsureApplicationServices may have failed due to a missing or throwing Startup class.
+                if (_applicationServices == null)
                 {
-                    addresses.Add("http://localhost:5000");
+                    _applicationServices = _applicationServiceCollection.BuildServiceProvider();
                 }
+
+                EnsureServer();
+
+                // Write errors to standard out so they can be retrieved when not in development mode.
+                Console.Out.WriteLine("Application startup exception: " + ex.ToString());
+                var logger = _applicationServices.GetRequiredService<ILogger<HostingEngine>>();
+                logger.LogError("Application startup exception", ex);
+
+                // Generate an HTML error page.
+                var runtimeEnv = _applicationServices.GetRequiredService<IRuntimeEnvironment>();
+                var hostingEnv = _applicationServices.GetRequiredService<IHostingEnvironment>();
+                var showDetailedErrors = hostingEnv.IsDevelopment()
+                    || string.Equals("true", _config[DetailedErrors], StringComparison.OrdinalIgnoreCase);
```

Comment:
on unix Hosting:DetailedErrors has to be set has `Hosting__DetailedErrors` as `:` isn't allowed in env vars.